### DISCUSSION TITLE
Mlife fatigue channel error

### DISCUSSION THREAD

**Discussion by lt72884** [OpenFAST/openfast#1745] [Q&A]
I successfully ran the Mlife cert tests as well as ran a basic "Test #01 baseline test" using the settingsFile.mlif that camew with the Test #01 baseline. 

I currently change the settings file in the "input files" section from DLC2_3.1.out to 5MW_Land_DLL_WTurb.out file that i generated using openFAST. I made sure i was NOT using the .outb file. 

The format of my .out file looks just like the format of the DLC file. 

In the settings file, i also made sure to change the column number from 50, and 51,  to match my column number as shown below:

FROM THIS:
"RootFMxy1" "(kN)"   "sqrt( timeSeriesData(:,50).^2 + timeSeriesData(:,51).^2 )"
"RootMMxy1" "(kN·m)" "sqrt( timeSeriesData(:,53).^2 + timeSeriesData(:,54).^2 )"

TO THIS:
"RootFMxy1" "(kN)"   "sqrt( timeSeriesData(:,30).^2 + timeSeriesData(:,31).^2 )"
"RootMMxy1" "(kN·m)" "sqrt( timeSeriesData(:,33).^2 + timeSeriesData(:,34).^2 )"

I run the following code and then get the following error:

settingsFile = 'settingsFile - Copy.mlif'
dataDirectory = 'G:\Cloud Master\UVU\Windmill internship\Mlife Testing\Source\'
outputDirectory = 'G:\Cloud Master\UVU\Windmill internship\Mlife Testing\Data Dump\'
[Fatigue, Statistics]= mlife( settingsFile, dataDirectory, outputDirectory );

Error using [process_mlife_inputs](matlab:matlab.internal.language.introspective.errorDocCallback('process_mlife_inputs'))
  For fatigue channel #1, the channel number must be between 1 and 54 (inclusive).

Error in [mlife](matlab:matlab.internal.language.introspective.errorDocCallback('mlife', 'G:\Cloud Master\UVU\Windmill internship\Mlife Testing\Source\mlife.m', 275)) ([line 275](matlab: opentoline('G:\Cloud Master\UVU\Windmill internship\Mlife Testing\Source\mlife.m',275,0)))
   [FileInfo, Statistics, Fatigue] = process_mlife_inputs(FileInfo, Statistics, Fatigue, dataDirectory);

thanks for any help and advice. 

Incase you need the last section of the settings file:
-----  Input Files  ------------------------------------------------------------
1                 FileFormat         Format of input files.  1 = FAST ascii, 2 = FAST binary
1  1.1   1.3   1.5   1.7    (Weibull-Weighted Normal Operation: NumNormFiles, PSF1, PSF2, PSF3, PSF4)
"5MW_Land_DLL_WTurb.out" 
0  1.1   1.3   1.5   1.7    (Weibull-Weighted Idling: NumIdleFiles, PSF1, PSF2, PSF3, PSF4)
0  1.2   1.3   1.4   1.6    (Discrete Events: NumDiscFiles, PSF1, PSF2, PSF3, PSF4)
==EOF==                             DO NOT REMOVE OR CHANGE.  MUST COME JUST AFTER LAST LINE OF VALID INPUT.

**Comment by jjonkman** [comment]
Dear @lt72884,

It looks like you changed the channel numbers in the CALCULATED CHANNELS section of the MLife input file.  Did you also change the channel numbers in the FATIGUE section?  I'm thinking specifically of the first column of the following table:
`Channel#  NSlopes  SNslopeLst  BinFlag  MaxBinWidth/Number  TypeLMF  LUlt`

The error implies that you have set `Channel#` to an index higher than the number of channels (54) in your OpenFAST output file.

Best regards,

**Comment by lt72884** [comment]
For your reference from the matlab ChanName output
'RootFxb1' = 29
'RootFyb1' =30
'RootFzb1' =31
'RootMxb1' =32
'RootMyb1' =33

As per your advice, i changed it to be the following. Is this correct?
Channel#  NSlopes  SNslopeLst      BinFlag  MaxBinWidth/Number  TypeLMF     
   29       1        10            BW		10.0      	175       1750
    30       1        10            BW		60.0      	309       3090
   32       1        10            BW		100.0     	5000      50000
    33       1        10            BW		10.0        	5        500

Lult and beyond is empty, do i need these values? is there a fully completed settings file to look at to test? I am trying to get cycles to failure. 

also, for the equivalent frequewncy in Hz, im using 2.11 since a FFT in matlab shows the peak freq to be 2.11, i changed it from the default 1 to 2.11

Thank you for the help:)

**Comment by lt72884** [comment]
hey, guess what, i got some output. It is difficult to read since it output it as a .txt file but i did get it to export to a excel file. Im super close to my goal of cycles to failure. will mlife do that?

thanks

**Reply by jjonkman** [reply]
MLife can export text or MS Excel files.  MLife can output short-term and lifetime damage equivalent loads, damage, and time until failure.

Best regards,

**Comment by lt72884** [comment]
Thank you much for all the help. Here is the output in excel. Is the time until failure in seconds, minutes, days, ? 
Do i need to add an Lult in the settings file since mine is blank or is that a value that is calculated?

I found this in the guide, but not sure where to find these values at? IE, how do i find my LMF and ultimate load to add to the settings file?

""Each fatigue channel requires a number of input parameters: associated Wohler 
exponent(s), load range binning values, the fixed-mean load, and the ultimate load. An 
example of this input table (green highlight) is shown in Figure 3.""
thanks

https://photos.app.goo.gl/615DNUME6wmcFL9r6

**Reply by jjonkman** [reply]
The time until failure is reported by MLife in seconds (s), as indicated.

**Comment by lt72884** [comment]
LOL, ok, i read that as failure or failures hahaha. Thanks fro clearign that up. 

What about the input parameters i asked about? such as binning values, LMF, Lut etc etc? 

thanks for all the help.

**Reply by jjonkman** [reply]
I'm not sure I understand your question about the values in the FATIGUE table.  It looks like you've specified all 7 values of the table.

**Comment by lt72884** [comment]
in the settings file, they are all empty, but the mlife user manual says i need to populate them? 
He is the snippet form the user manual, plus a picture of my settings file.

"Each fatigue channel requires a number of input parameters: associated Wohler
exponent(s), load range binning values, the fixed-mean load, and the ultimate load. "

https://photos.app.goo.gl/jDLgzberLbYV8vAz6

thanks

**Reply by jjonkman** [reply]
The table has 7 columns and I see that you've entered 7 columns of data, although your data are not aligned with the column headers.

**Comment by lt72884** [comment]
ok, i see what you are saying. thanks for showing me that. However, i did not enter those values. Did they auto populate or are they from someone else? i am using "Test01" settings file. They must be in there by default? If so, they may be the wrong values? If they are the wrong values, how do i add the correct values? 

thank you:)

**Reply by jjonkman** [reply]
It looks like you are using the values from _Test01.mlif_ in the MLife CertTest.  The most challenging inputs to define are `TypeLMF` and `LUlt`.  The use of MLife and the setting of `TypeLMF` and `LUlt` have been been discussed many times on our forum: https://forums.nrel.gov/. I would first search there. We have developed a FAQ specifically around `TypeLMF` and `LUlt`, as included in my post dated Jun 14, 2020 in the following forum topic: https://forums.nrel.gov/t/mlife-fatigue-del-calculation/661.

**Comment by lt72884** [comment]
thank you for the reply. I have seen links to the faq, but none of them take me there. They all take me to the homepage of NREL

https://www.nrel.gov/wind/nwtc.html

so i am unable to find the FAQ. I sent in a request the other day via the "contact webmaster" button on the site to try and find the correct links or pages. 

here is a picture of the page the link takes me too.

https://photos.app.goo.gl/5wgNcAnGbhM7C3vq7

I need to re-register for the forums because i forgot to send the email to you haha

**Reply by jjonkman** [reply]
The old FAQ was on the old NWTC Information Portal website that crashed, and is no longer available.  However, I posted the answer from the old FAQ regarding `TypeLMF` and `LUlt` in my post dated Jun 14, 2020 in the forum topic I linked to above.

**Comment by lt72884** [comment]
oh, ok, ill look through the thread again and see your reply on lmf and lult

**Comment by lt72884** [comment]
ok, great i found the section, changed my LMF to AM. I am now calculating the Lult based off of a table i have from a book. Now if i am reading the FAQ post correctly, Lult is based on strength of the cross sectional area, units should be MPa since its F/A? 

second LMF is also a stress? IE, the LMF is sigma mean (sigma m) of a cyclic fatigue wave as shown in the image below?

https://media.springernature.com/lw685/springer-static/image/prt%3A978-0-387-92897-5%2F6/MediaObjects/978-0-387-92897-5_6_Part_Fig1-1176_HTML.gif


thanks

**Reply by jjonkman** [reply]
The units of TypeLMF and LUlt should be the same as the units of the load channel you are post-processing.

**Comment by lt72884** [comment]
ok, thanks. 

since i do not have FEA available, when i look at pydatview of RootFxb1 [kN], it shows the MAX is 444kN. Is this the going to be the same overall maximum load as MExtremes would find?

Here is a picture of pydatview
https://photos.app.goo.gl/CPr1gcrsR4q47xY36

here is the FAQ quote im referring too.

. If this is not available—for conceptual designs for example—what we’ve done in the past is to see how the fatigue loads vary over a range of LUlt values. We’ve defined a range by taking the overall maximum value of the load in question (e.g., derived from MExtremes), 

thanks

**Reply by jjonkman** [reply]
Dear @lt72884,

MExtremes can find the maximum from specific time series, but it does many other things such as calculate the extremes from a large number of time series, compute extreme event tables, apply scales, offsets, and partial safety factors, identify peaks beyond what is in the data through a peak finding algorithm, etc.

Best regards,

**Comment by lt72884** [comment]
Ok, thanks for that information.

Now, when I view rootfxb1 with pydatview and it says it's max load for that channel is 444kN, is that the maximum value of that load that is needed to now multiply by the scalers in the faq? Ie 1.25, 1.75, 2, 4, 10, 20 etc?

I don't have MExtremes yet so I'm hoping the 444kn max value to use at the moment 

Thanks

**Reply by jjonkman** [reply]
Have you run many OpenFAST simulations across a range of conditions and computed the maximum value from all simulations?  This is the value referred to in the FAQ.

Best regards

**Comment by lt72884** [comment]
Yes, I have about 75 open fast files with different wind conditions etc.

Is the maximum value you speak of, the one pydatview is showing? Ie, the max value of the time series data of that channel? 

Thanks

**Reply by jjonkman** [reply]
pyDatView can compute the maximum value from each simulation that is selected, but I don't believe it does what MExtremes does to identify the extremes across all output files.  From the image you shared, though, it looks like you were only processing one time series in pyDatView.

Best regards,

**Comment by lt72884** [comment]
yeah, it was the time series of channel 29 (RootFxb1)  in my case. So out of 48001 time steps, the max value is 444KN. I did get MExtremes working and its the same value, however, MExtremes has more information. 

What i need help with next is understanding the data Mlife just produced ;

I used 444kN*10 as my ultimate load based on the information from MExtremes, and cycles to failure in seconds is 2.26*10^16 seconds which is thousands of years haha. Im guessing infinite life? 

i may create another thread for hte help. 

BUT should i do that here on github or go create an account on the forums? 

thanks

**Reply by jjonkman** [reply]
We use the Issues section on GitHub to track bugs, new features, etc.  We use the Discussion section on GitHub similar to posts on our forums (https://forums.nrel.gov/); that is, for general user-related support.  IThere have been many, many posts on our forums related to OpenFAST modeling over the years, so, I would always search there before posting new questions on GitHub or on the forum; it is likely that your question has been asked and answered before.

**Comment by lt72884** [comment]
awesome sounds great. I just tried making a turbsim file with 24 hours of useable time.... now its a 10GB file haha. I think 24 hours is to much data. might just do an hour or so hahaha

**Comment by lt72884** [comment]
ok, i have another question. On the balde root for the 5mw dll land folder, on the physical blade itself, how many threaded studs/bolts are used to attach it to the hub? whats size are they? i have read values from m-45 to m-90 and from 54 to 120 of them. 

thanks

**Reply by jjonkman** [reply]
Dear @lt72884,

These details regarding root bolts have not been specified for the NREL 5-MW baseline wind turbine blade.  You'll have to make your own assumptions for your own purposes.

Best regards,

**Comment by lt72884** [comment]
ok, sounds good. thanks. I do know they use m-45 grade 10.9 threaded studs that can easily handle 940MPa which is by far higher than the ultimate load these wind turbines experience even under stressful situations. Those bolts could handle hurricane weather haha.

My max load from the simulation is 666kN which is not to bad.

**Comment by lt72884** [comment]
is openfast capable of adding cracks and defects to the blade to test failure?

thanks

**Reply by jjonkman** [reply]
Dear @lt72884,

See a similar question asked and answered on our forum: https://forums.nrel.gov/t/making-fault/1260.

Best regards,

**Comment by lt72884** [comment]
ok thank you. 

Where can i find the equations openfast uses to calculate RootMxc1 and all of the outlists i have added? i am curious to see how openfast calculates things, especially moments and forces. 

thanks

**Reply by jjonkman** [reply]
See my post dated Apr 29, 2010 in the following forum topic for an explanation of how the ElastoDyn module of OpenFAST calculates the blade-root moments (via load summation): https://forums.nrel.gov/t/tower-base-axial-force-discrepancy-between-fast-and-bladed/238.

**Reply by lt72884** [reply]
fabulous. thanks for this. I have learned so much the last few months. thanks for sticking it out with me and helping.

is there a way i can see the code of the equations?

**Comment by lt72884** [comment]
is there by chance a cad file of the blade used for the 5MW lan dll wturb?

**Reply by jjonkman** [reply]
See my post dated Aug 29, 2023 in the following topic on our forum for my response to a similar question about the availability of a surface mesh for the NREL 5-MW baseline wind turbine: https://forums.nrel.gov/t/nrel-5-mw-blade-airfoils-data-files-ar17/5561.

Best regards,

**Comment by lt72884** [comment]
thank you :)
